Expose serialized template content on WP_Block_Template.

nerrad opened the pull request:
When building the template, initialize WP_Block_Template with the serialized block content so callbacks registered to the `hooked_block_types` filter can access the content.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

If you registered a callback to `hooked_block_types` when the callback receives an instance of a template or template_part via the `$context` argument, the `$content` property will be null. Minimally reproducible example:

```php
<?php
add_action( 'hooked_block_types', 'testing_auto_hook', 10, 4 );
function testing_auto_hook( $hooked_blocks, $position, $anchor_block, $context ) {
   if ( $context instanceof \WP_Block_Template ) {
	$content = $context instanceof \WP_Block_Template && $context->content ? $context->content : '';
	if ( strpos( $content, 'wp:post-title' ) !== false ) {
		// the `core/post-title` block already exists in content so abort.
		return $hooked_blocks;
	}
	if (
		$context instanceof \WP_Block_Template &&
		'after' === $position &&
		'core/post-title' === $anchor_block
	) {
		$hooked_blocks[] = 'core/navigation';
	}
   }
   return $hooked_blocks;
}
```

In the above example, the code is checking to see if the `$context` is an instance of `WP_Block_Template` and if it is, looks in the `$content` property for the serialized representation of the `core/post-title` block. If it's present, then bail early and don't inject navigation. For the pages template, I'd _expect_ this to not inject `core/navigation` because the `core/post-title` block is present in the content, but in fact, it does get injected because `WP_Block_Template::$content === null`.

Screenshots:

| Before | After (with this PR/patch) |
| ----- | ------ |
| ![CleanShot 2023-11-10 at 14 50 39@2x](https://github.com/nerrad/wordpress-develop/assets/1429108/4958e2f9-f67b-435c-9cd0-c06bef472b0d) |  ![CleanShot 2023-11-10 at 14 50 13@2x](https://github.com/nerrad/wordpress-develop/assets/1429108/b6c78b55-f93b-431c-bcae-7eaa8739f54e) |


Trac ticket: https://core.trac.wordpress.org/ticket/59882

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

nerrad committed:
Expose serialized template content on WP_Block_Template.

When building the template, initialize WP_Block_Template with the serialized block content so callbacks registered to the `hooked_block_types` filter can access the content.

ockham commented:
Makes sense, thank you for your contribution!

If we do this, should we also drop the `$template_content` variable altogether? It's currently [parsed for the actual hooked blocks insertion](https://github.com/WordPress/wordpress-develop/blob/00791300b386af12d871293c56cd8a2e208f4b32/src/wp-includes/block-template-utils.php#L557-L558); it might be less confusing for people if we changed that to also used `$template->content`.

nerrad committed:
Remove unnecessary variable 

Consolidates initial read from template file into the `content` property in the `WP_Block_Template` instance removing the need for a separate `$template_content` variable.

nerrad commented:
Yup makes sense 👍🏻

nerrad commented:
One thing to be aware of with this change is that it might appear to extenders that it's possible to mutate `$template->content`. The blocks are already parsed from content by this point but one extension mutating the property could mess it up for every other callback utilizing block_hooks and checking this property (the content wouldn't match the rendered blocks). Ideally, $template->content would be a private property exposed via getter so that extenders can't mutate it.

ockham commented:
Yeah, good point 🤔 

Just to make sure, are you still okay if we proceed (with https://github.com/WordPress/wordpress-develop/pull/5656/commits/9eebbbed451cd63ac52f3a04ad52dfd5df6a3b20)? Personally, I'm inclined to say that it's okay to expect an extender _not_ to mess up `$template->content`, or at least I don't think it's a blocker.

Also agree that a getter would be nice to have for the template content, but OTOH, we're also [overriding it with new markup returned from `traverse_and_serialize_blocks`](https://github.com/WordPress/wordpress-develop/pull/5656/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7R558) (with hooked blocks inserted) so I'm not sure we can make the `content` property `private`, as [PHP doesn't seem to have a `friend` concept like C++ does](https://stackoverflow.com/questions/317835/php-equivalent-of-friend-or-internal) 🤔 

Anyway, mostly just curious if you think this PR is blocked by any of the above 😄

nerrad commented:
Yup, I'm okay with the latest commit. I just wanted to flag it as a potential issue to get further thoughts. The only way (I can currently think of) that we'd be able to lock this down is to require a new object constructed and use something the [value object pattern](https://dzone.com/articles/practical-php-patterns/basic/practical-php-patterns-value) to keep `WP_Block_Template` immutable. Practically, I'm not sure how big of a deal that is.

ockham commented on src/wp-includes/block-template-utils.php:
```diff
@@ -532,6 +531,7 @@ function _build_block_template_result_from_file( $template_file, $template_type
 	$template->has_theme_file = true;
 	$template->is_custom      = true;
 	$template->modified       = null;
+	$template->content        = file_get_contents( $template_file['path'] );
```
One final suggestion: Let's set `$template->content` right below `$template->theme` (i.e. on [`R526`](https://github.com/WordPress/wordpress-develop/pull/5656/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7R526) if I'm not mistaken).

That's the [position](https://github.com/WordPress/wordpress-develop/pull/5192/files#diff-1d8e05d3ea69557ca8742330e62c1893a72d4ca4ff2ed4ccc5b94bcf5d20fed7L568) where it originally was before we started porting Block Hooks code to Core, and it's also more consistent with `_build_block_template_result_from_post`: https://github.com/WordPress/wordpress-develop/blob/0b8ca16ea3bd9722bd1a38f8ab68901506b1a0e7/src/wp-includes/block-template-utils.php#L760-L762

ockham reviewed (commented):
One final [note](https://github.com/WordPress/wordpress-develop/pull/5656/files#r1395743713); otherwise LGTM!

ockham commented:
I've taken the liberty of [moving the line](https://github.com/WordPress/wordpress-develop/pull/5656#discussion_r1395743713) and committed this change to Core: https://core.trac.wordpress.org/changeset/57118.

Hope that's okay! 😊

ockham commented:
Backported to the 6.4 branch in https://core.trac.wordpress.org/changeset/57119.
